### DISCUSSION TITLE
Help with caption, I'm unable to wrap text when it's too close to border

### DISCUSSION THREAD

**Discussion by bruno-smaldone** [ImageMagick/ImageMagick#3389] [Help]
Hi there,

I found this while trying to overlay wrapped text on an image.

Here's an example showing the expected behavior. The strings 'aaaaa', 'bbbbb', 'ccccc' and 'ddddd' were moved to the line below because they fit in the same line, inside the defined image space.

```
convert "-size" "250x250" "-fill" "black" "-pointsize" "40" "caption:aaaaaaaa bbbbbbbb cccccccc dddddddd " "./test.jpg"
```

Result:
![first_test](https://d2vxgxvhgubbj8.cloudfront.net/custom/7698/IM_01.jpg)

Now, if we increase the length of one of those words, the wrapping stops working as expected:

```
convert "-size" "250x250" "-fill" "black" "-pointsize" "40" "caption:aaaaaaaa bbbbbbbbbbbb cccccccc dddddddd" "./test.jpg"
```
Result:
![second_test](https://d2vxgxvhgubbj8.cloudfront.net/custom/7698/IM_02.jpg)

See how there's no jump line between 'bbbbbbbbbbbb' and 'cccccccc' even when 'cccccccc' is clearly outside of the image dimensions.

Now, if we further increase the size of the second word, then the wrapping works fine again.

```
convert "-size" "250x250" "-fill" "black" "-pointsize" "40" "caption:aaaaaaaa bbbbbbbbbbbbbbbb cccccccc dddddddd " "./test.jpg"
```
Result:
![third_test](https://d2vxgxvhgubbj8.cloudfront.net/custom/7698/IM_03.jpg)

Do you know if there's any workaround I can apply here or if I am doing something wrong?

Any help is appreciated.

Thank you,

**Config:**

```
Version: ImageMagick 7.0.10-34 Q16 x86_64 2020-10-09 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5) 
Delegates (built-in): bzlib freetype gslib heic jng jp2 jpeg lcms lqr ltdl lzma openexr png ps tiff webp xml zlib
```
BigSur 11.1 (20C69)

**Comment by fmw42** [comment]
If I understand your question, that is the way wrapping works, though there should not be a partial word. But  you can insert \n where you want to force a new line so that the next set of letters start on a new line. That works best with label: rather than caption:.  The IM developers will need to look into the partial letter/word?

**Reply by bruno-smaldone** [reply]
Hi @fmw42 , thanks for the quick comment.

The problem I'm seeing is that the second example's command is not wrapping the text. That's causing 'cccccccc' to not be visible.

```
convert "-size" "250x250" "-fill" "black" "-pointsize" "40" "caption:aaaaaaaa bbbbbbbbbbbb cccccccc dddddddd" "./test.jpg"
```

![second_test](https://d2vxgxvhgubbj8.cloudfront.net/custom/7698/IM_02.jpg)

The expected behavior would be, in my opinion, that 'cccccccc' should be visible in the third line, because 'bbbbbbbbbbbb' and 'cccccccc' together are too wide to live in the same line. Does that make sense?

Sorry if my explanation is not very clear, second language 🤓 I'd be happy to answer questions or add more examples, just let me know!

Thanks,

PS. I'm using text-wrapping to avoid using "manual jump lines" (\n), since the idea is to create images programmatically and unsupervised.

**Reply by fmw42** [reply]
I will have to leave this to the IM developers. But if the font is such that the image is shorter than needed for the last line to wrap, it will not show. There are 3 variables to consider in wrapping: W, H and pointsize. If you want it to wrap, use only two. For example it works for me, if you remove the pointsize.

**Reply by bruno-smaldone** [reply]
Thanks again for your input Fred, I really appreciate it.

From a user perspective, there're some trade-offs to that, which I will comment on below as I think it contributes to the conversation.

- If you wrap based on image width and height, it means the pointsize will be automatically adjusted, and you will lose control of how big the characters should be

- If the wrap is based on image width and point size, the words are cut in half instead of overflowing when they are too big, adding an example below:

```
convert "-size" "250x" "-fill" "black" "-pointsize" "40" "caption:aaaaaaaa bbbbbbbbbbbbbbbb cccccccc dddddddd " "./test.jpg"
```

Result:

![second_test](https://d2vxgxvhgubbj8.cloudfront.net/custom/7698/IM_04.jpg)

Thanks again for your comment! 

I'm looking forward to the dev comments to gain more info on this

**Comment by snibgo** [comment]
"caption:" should insert line-breaks between words to prevent the word 'cccccccc' from entirely disappearing.

The problem occurs in v7.0.8-64 and v7.0.10-45. It does not occur in v6.9.9-50.

This seems to be a bug.

**Comment by urban-warrior** [comment]
If you specify both the pointsize and size of the canvas, caption's auto fitting is disabled.  However, we'll add support for a new define, `-define caption:split=true` in the next point release to return expected results.

**Reply by bruno-smaldone** [reply]
That sounds terrific, thanks for the answer!
